java/core/src/test/java/org/bondlib/protocol/CompactBinaryWriterTest.java

```diff
@@ -111,4 +119,46 @@ public void testFirstPassWriterWriteFieldOmittedDoesNotWrite() {
             fail("IOException can't be thrown here: " + e);
         }
     }
+
+    /*
+     * Regression test for a bug where there would be stale state after
+     * serializing a Bonded field with a Compact v2 writer (the only stateful
+     * writer). The bug was fixed in commit 8a608fbe.
+     */
+    @Test
+    public void v2WReuseWithBonded() throws IOException {
+        final int objCount = 3;
+
+        final List<Base> basesIn = new ArrayList<Base>();
+        for (int i = 0; i < objCount; i++) {
+            final Base base = new Base();
+            base.baseInt = i;
+            basesIn.add(base);
+        }
+
+        final List<BondedBase> bondedBasesIn = new ArrayList<BondedBase>();
+        for (int i = 0; i < objCount; i++) {
+            final BondedBase bondedBase = new BondedBase();
+            bondedBase.bondedBase = Bonded.fromObject(basesIn.get(i), Base.BOND_TYPE);
+            bondedBasesIn.add(bondedBase);
+        }
+
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        final CompactBinaryWriter writer = new CompactBinaryWriter(baos, 2);
+        final Serializer<BondedBase> serializer = new Serializer<BondedBase>();
+        for (int i = 0; i < objCount; i++) {
+            serializer.serialize(bondedBasesIn.get(i), writer);
+        }
+
+        final ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
+        final CompactBinaryReader reader = new CompactBinaryReader(bais, 2);
+        final Deserializer<BondedBase> deserializer = new Deserializer<BondedBase>(BondedBase.BOND_TYPE);
+        final List<Base> basesOut = new ArrayList<Base>();
+        for (int i = 0; i < objCount; i++) {
```

Comment:
Consider reading until there's no more data in the input stream. This will also ensure that we successfully read everything that was written.

If you want to assert that there's nothing left to read, that would work too.